Early alpha android build

Burge337 opened the pull request:
This branch gets most of the core functionality of Storybuilder working on android devices, while additionally implementing logging for submitted sentences in the mongo database.

The only feature that is currently broken in the android build branch is the file selector for images when creating pages for a story. It relied on a Unity editor feature that can't exist in any actual build of the project.

Burge337 committed:
Create initial SUPER rough test build for storybuilder

This creates a version of storybuilder that will actually build for android devices. In making storybuilder build properly certain functionalities needed to be temporarily disabled until they were refactored. 

Non-comprehensive list of changes:
1) loading context packs
2) file explorer opening in pages with images users supply
3) server contact disabled (no logging enabled)

Burge337 committed:
Refactor local storage of context packs/etc and reimplement server calls

This swaps the storage location of LearnerData/ContextPacks/LeanerIcons to Application.persistentDataPath, which should work for IOS/Android/Mac/Windows as Unity changes the storage path based on the OS.

Also made a bunch of the buttons a lot bigger; using the app on a phone was quite painful with how small things were before. Context pack filters and letter filters for the wordbank should also be increased significantly.

Additionally, the end session times for LearnerData are now logged upon changing learners or logging out, as OnApplicationQuit() never seemed to fire from initial testing.

Burge337 committed:
Re-implement/fix sentence storage in storybuilder scene

Previously sentences weren't being stored in the Application.persistentDataPath and were thus not being used in anything but PC OS's. This changes their storage location to that persistent path/Resources/Sentences. No communication with the server occurs for sentences; storage is entirely local.

This also fixes an issue where sentences weren't actually updating in the sentence bank when we swapped to the storybuilder canvas. This was because the sentence bank population logic only occurred upon initialization, which only occurs once now that we've combined the storybuilder/sentencebuilder scenes into one. We now rely on OnEnable/OnDisable, where sentencebuilder => storybuilder triggers OnEnable and vice versa. Current implementation repopulates this sentence bank each time we re-load it rather than ensuring duplicate sentences aren't added, which may be something to consider re-factoring in the near future.

Burge337 committed:
Fix word tile submission not properly deleting tiles for android builds

This has now been refactored to simply change the parent of the wordtiles upon submission rather than copy and delete them. In order to keep the wordtiles from displaying over top of the sentence tile, they are also deactivated upon changing parents to sentenceInTiles, which then activates them when resubmitting sentences.

Also adjusted cell size/spacing for leaner buttons and removed the old and unused ClearTiles() method from SentenceBar.cs

Burge337 committed:
Remove local file storage of learnerdata

All learnerdata has been removed from local file storage to prevent redundant function calls. As this is purely logging data  for research purposes, we've decided that it's not something that the system user needs to have access to.

Burge337 committed:
Add error message to user login screen for failed logins

The UserLogin error message implementation is a bit weird in that it relies on updating a boolean loginFailed variable upon failing to login that Unity's Update() method then checks for each frame. This is because the login method takes place outside of the main thread, where the Unity API is not available. This could probably be made more efficient by checking less frequently in the future.

This also removes some old sentences we'd had stored that aren't accessible in android builds of the project.

Burge337 committed:
Fix saved sentences in storybuilder not preserving word form changes

SavedSentence's were refactored to include a string of the sentence text and a list of the word strings that make up said sentence. This was to get around the issue of being unable to change word.word without also changing all other word tiles with the same script attached.

Also changed the serialization for sentences from binary to regular old json files.

Burge337 committed:
Initial implementation of sentence server request POST/GET methods

This allows us to track a learner's submitted sentences, the time they were submitted, and the userId of the user that they belong to. 

A weird quirk of this implementation is that the actual server call that gets a learner's sentences from the database on startup is attached to the ChangeScene.cs script, where it would make more sense to have it attached to the SavedSentenceBank.cs script. This is because Unity's Start() method won't fire on SavedSentenceBank.cs until we've changed scenes (it starts disabled). This is a problem because changing scenes prompts loading sentences from local files which could then cause conflicting read/write operations or even weird race condition bugs.

TODO: add used context pack information to the SavedSentence class. This would help track which context packs a learner most interacted with.

Burge337 committed:
Refactor sentence submission to track context packs used

This entirely removes the ServerContextPack class and instead combines its fields with the older ContextPack class. This combination originally seemed like it would be a problem because of serialization issues/not wanting to track certain fields, but simply adding the [NonSerialized] tag to fields allows the class to deviate a bit from its server equivalent.

In this implementation the Word class now knows what contextPackId it's from in unity and the web client, without storing that information in the database. Instead, words simply hold their context pack Id for sentences to be able to figure out what context packs were used in their creation.

Burge337 committed:
Refactor sentence submission to track context packs used

This entirely removes the ServerContextPack class and instead combines its fields with the older ContextPack class. This combination originally seemed like it would be a problem because of serialization issues/not wanting to track certain fields, but simply adding the [NonSerialized] tag to fields allows the class to deviate a bit from its server equivalent.

In this implementation the Word class now knows what contextPackId it's from in unity and the web client, without having the words of a submitted sentence in the database also keep track of their contextPackId. (there is just the one list of context packs involved in the sentences creation)

Burge337 committed:
Merge remote-tracking branch 'origin/early-alpha-android-build' into early-alpha-android-build

kklamberty reviewed (commented):
@Burge337 - Can you resolve the conflicts (in GitKracken is probably fine like we did before) so that I know what to run on my machine to test this branch? Thanks!

Burge337 committed:
Merge branch 'master' into early-alpha-android-build

Burge337 committed:
Remove unused method

Burge337 committed:
Refactor database words to no longer contain contextPackId's

This was changed because the previous implementation was hugely redundant in storing context pack Id's. The new implementation only lets storybuilder words/word tiles know their context Id's to remove the aforementioned mongo database redundancy.
